File: tests/matlab/short_retrain_2lpt.m
function short_retrain_2lpt()
% Step A.3 MATLAB lane. Runs minFunc/L-BFGS for ~50 iterations on the
% same 1300-spectrum 2lpt training set the Python A.3 runner uses,
% via the legacy MATLAB objective.m (which has correct zqso_1pz +
% BOSS DR12Q priors at lines 47, 66-77).
%
% Saves <fixture>/short_retrain/matlab.mat with final M, log_omega,
% log_c_0, log_tau_0, log_beta + loss trajectory captured via a global.
%
% Run via:
%   /sw/pkgs/arc/matlab/R2024b/bin/matlab -batch \
%     "addpath(fullfile(pwd,'tests','matlab')); short_retrain_2lpt"

  this_file = mfilename('fullpath');
  this_dir  = fileparts(this_file);
  fixture_dir = fullfile(this_dir, '..', 'fixtures', '2lpt_frozen');
  out_dir = fullfile(fixture_dir, 'short_retrain');
  if ~exist(out_dir, 'dir'); mkdir(out_dir); end

  % Setup: legacy MATLAB code + minFunc
  matlab_repo = '/home/mfho/MATLAB/gp_dla_detection_dr16q_public';
  addpath(matlab_repo);
  addpath(genpath('/home/mfho/MATLAB/minFunc_2012'));
  if isempty(which('minFunc'))
    error('minFunc not on path');
  end
  if isempty(which('objective'))
    error('legacy objective.m not on path');
  end

  init = load(fullfile(fixture_dir, 'init_params.mat'));
  train = load(fullfile(fixture_dir, 'training_set.mat'));

  M0 = double(init.M);
  log_omega0 = double(init.log_omega(:));
  log_c_0_0 = log(double(init.c_0(1)));
  log_tau_0_0 = log(double(init.tau_0(1)));
  log_beta_0  = log(double(init.beta(1)));

  centered_fluxes = double(train.centered_fluxes);
  noise_variances = double(train.noise_variances);
  z_qsos          = double(train.z_qsos(:));
  rest_wavelengths = double(init.rest_wavelengths(:));

  % Replace stored NaN-fluxes with NaN (already there in fixture); MATLAB
  % objective.m uses ~isnan(...) for masking
  N = size(centered_fluxes, 1);
  n_pix = size(centered_fluxes, 2);
  k = size(M0, 2);

  % lya_1pzs (N × n_pix) — same formula Python uses
  lya_rest = 1215.67;
  lya_1pzs = (1 + z_qsos) * (rest_wavelengths' / lya_rest);

  num_forest_lines = double(init.num_forest_lines(1));
  TW = double(init.all_transition_wavelengths(:));
  OS = double(init.all_oscillator_strengths(:));

  % Build initial x packing required by legacy objective.m:
  %   x = [vec(M); log_omega; log_c_0; log_tau_0; log_beta]
  initial_x = [M0(:); log_omega0; log_c_0_0; log_tau_0_0; log_beta_0];
  expected_len = n_pix * k + n_pix + 3;
  if length(initial_x) ~= expected_len
    error('initial_x length %d != expected %d', length(initial_x), expected_len);
  end
  fprintf('initial_x: %d entries (n_pix=%d, k=%d)\n', length(initial_x), n_pix, k);

  % Track loss trajectory via a persistent global (minFunc doesn't expose
  % a per-iter callback hook directly).
  global SHORT_RETRAIN_HIST;
  SHORT_RETRAIN_HIST = struct( ...
    'loss',     [], ...
    'log_c_0',  [], ...
    'log_tau_0',[], ...
    'log_beta', []);

  fwrap = @(x) wrapped_obj(x, centered_fluxes, lya_1pzs, noise_variances, ...
                           num_forest_lines, TW, OS, z_qsos);

  options = struct( ...
    'MaxIter',     50, ...
    'MaxFunEvals', 100, ...
    'Display',     'iter', ...
    'optTol',      1e-6, ...
    'progTol',     1e-9);

  fprintf('Starting minFunc...\n');
  t0 = tic;
  [x_final, fval_final, exit_flag, output] = minFunc(fwrap, initial_x, options);
  wall = toc(t0);
  fprintf('minFunc done: exit_flag=%d, iter=%d, funcCount=%d, fval=%.4f, wall=%.1fs\n', ...
    exit_flag, output.iterations, output.funcCount, fval_final, wall);

  % Unpack final params
  M_final          = reshape(x_final(1:n_pix*k), [n_pix, k]);
  log_omega_final  = x_final(n_pix*k+1 : n_pix*(k+1));
  log_c_0_final    = x_final(end-2);
  log_tau_0_final  = x_final(end-1);
  log_beta_final   = x_final(end);

  % Save
  out = struct();
  out.lane             = 'matlab';
  out.M_final          = M_final;
  out.mu               = double(init.mu);
  out.log_omega_final  = log_omega_final;
  out.log_c_0_final    = log_c_0_final;
  out.log_tau_0_final  = log_tau_0_final;
  out.log_beta_final   = log_beta_final;
  out.c_0_final        = exp(log_c_0_final);
  out.tau_0_final      = exp(log_tau_0_final);
  out.beta_final       = exp(log_beta_final);
  out.loss_history     = SHORT_RETRAIN_HIST.loss(:);
  out.log_c_0_history  = SHORT_RETRAIN_HIST.log_c_0(:);
  out.log_tau_0_history = SHORT_RETRAIN_HIST.log_tau_0(:);
  out.log_beta_history  = SHORT_RETRAIN_HIST.log_beta(:);
  out.rest_wavelengths = rest_wavelengths;
  out.exit_flag        = exit_flag;
  out.iterations       = output.iterations;
  out.funcCount        = output.funcCount;
  out.wall_s           = wall;

  save_path = fullfile(out_dir, 'matlab.mat');
  save(save_path, '-struct', 'out');
  fprintf('saved: %s\n', save_path);
end


function [f, g] = wrapped_obj(x, cf, lp, nv, nfl, tw, os, zq)
% Wrapper around legacy objective.m that records (loss, log_c_0,
% log_tau_0, log_beta) per call into a global for trajectory plots.
  [f, g] = objective(x, cf, lp, nv, nfl, tw, os, zq);
  global SHORT_RETRAIN_HIST;
  SHORT_RETRAIN_HIST.loss(end+1)      = f;
  SHORT_RETRAIN_HIST.log_c_0(end+1)   = x(end-2);
  SHORT_RETRAIN_HIST.log_tau_0(end+1) = x(end-1);
  SHORT_RETRAIN_HIST.log_beta(end+1)  = x(end);
end

File: tests/short_retrain_2lpt.py
"""Step A.3 — short retrain on the 1300-spectrum 2lpt fixture.

Runs full-batch Adam for N_ITERS=50 iterations on:

  v1 lane    : gpy_dla_detection.objective.spectrum_loss            (approx dlog_β)
  v3.5 lane  : gpy_dla_detection.training_v3_5.objective.spectrum_loss  (strict dlog_β)

Both lanes:
  - use the same starting parameters from tests/fixtures/2lpt_frozen/init_params.npz
  - use the same training data from tests/fixtures/2lpt_frozen/training_set.npz
  - bypass v1's `objective` wrapper (which has a separate zqso_1pz bug —
    see docs/notes/2026-05-07_v1_objective_zqso_bug_finding.md). Each
    spectrum is fed `zqso_1pz = z_qso + 1` correctly.
  - apply the BOSS DR12Q Gaussian priors on (log_τ_0, log_β) matching
    MATLAB legacy defaults, so the converged β is directly comparable
    between Python and MATLAB lanes.
    τ_0 ~ N(0.00554, 0.00064²)
    β   ~ N(3.182, 0.074²)

Saves to tests/fixtures/2lpt_frozen/short_retrain/<lane>.npz:
  loss_history, log_tau_0_history, log_beta_history, log_c_0_history,
  M_final, mu, log_omega_final, log_c_0_final, log_tau_0_final, log_beta_final.

Run:
    python tests/short_retrain_2lpt.py [--lane v1 | v3.5 | both]
                                       [--n-iters 50]
                                       [--lr 0.01]
"""
from __future__ import annotations

# 2026-05-07 PERFORMANCE NOTE: this trainer iterates over 1300 spectra
# inside Python (no batch vectorization yet — Step B). Each call to
# spectrum_loss does small linear-algebra ops (a 30×30 Cholesky + (n×30)
# matmuls). With multi-threaded BLAS (default OMP_NUM_THREADS=#cores),
# every inner-loop iteration spawns a thread storm whose synchronization
# overhead dwarfs the actual compute. Result: per-iter time blows up
# 10–20× under default threading.
#
# Pinning to 1 thread makes the inner loop 10× faster on this machine.
# Step B (vectorize across spectra) will let us re-enable multi-threaded
# BLAS profitably.
import os as _os
for _name in ("OMP_NUM_THREADS", "MKL_NUM_THREADS", "OPENBLAS_NUM_THREADS"):
    _os.environ.setdefault(_name, "1")

import argparse
import importlib
import sys
import time
from pathlib import Path

import numpy as np
import torch
import torch.nn as nn

sys.path.insert(0, str(Path(__file__).resolve().parent.parent))

FIX = Path(__file__).resolve().parent / "fixtures" / "2lpt_frozen"
OUT = FIX / "short_retrain"
DTYPE = torch.float64

# BOSS DR12Q priors (matches MATLAB objective.m:66-77)
TAU_0_PRIOR_MU = 0.00554
TAU_0_PRIOR_SIGMA = 0.00064
BETA_PRIOR_MU = 3.182
BETA_PRIOR_SIGMA = 0.074


def _import_spectrum_loss(lane):
    if lane == "v1":
        return importlib.import_module("gpy_dla_detection.objective").spectrum_loss
    if lane == "v3.5":
        return importlib.import_module("gpy_dla_detection.training_v3_5.objective").spectrum_loss
    raise ValueError(lane)


def _load_fixture():
    init = np.load(FIX / "init_params.npz")
    train = np.load(FIX / "training_set.npz")
    return init, train


def _build_init_params(init):
    """Returns (mu, M_param, log_omega_param, log_c_0_param,
                 log_tau_0_param, log_beta_param, TW, OS, num_forest_lines).
    Parameters are nn.Parameter so optimizer can update."""
    mu = torch.tensor(init["mu"], dtype=DTYPE)  # NOT optimized (kept fixed at init)
    M = nn.Parameter(torch.tensor(init["M"], dtype=DTYPE))
    log_omega = nn.Parameter(torch.tensor(init["log_omega"], dtype=DTYPE))
    log_c_0 = nn.Parameter(torch.tensor(np.log(float(init["c_0"])), dtype=DTYPE))
    log_tau_0 = nn.Parameter(torch.tensor(np.log(float(init["tau_0"])), dtype=DTYPE))
    log_beta = nn.Parameter(torch.tensor(np.log(float(init["beta"])), dtype=DTYPE))
    TW = torch.tensor(init["all_transition_wavelengths"], dtype=DTYPE)
    OS = torch.tensor(init["all_oscillator_strengths"], dtype=DTYPE)
    num_forest_lines = int(init["num_forest_lines"])
    return mu, M, log_omega, log_c_0, log_tau_0, log_beta, TW, OS, num_forest_lines


def _compute_lya_1pzs(rest, z_qsos, lya_rest=1215.67):
    """rest: (n_pix,); z_qsos: (N,) → returns lya_1pzs (N, n_pix)."""
    rest_t = torch.tensor(rest, dtype=DTYPE)
    z = torch.tensor(z_qsos, dtype=DTYPE)
    return (1.0 + z[:, None]) * rest_t[None, :] / lya_rest


def _full_batch_objective(spectrum_loss, mu, M, log_omega, log_c_0, log_tau_0, log_beta,
                          centered_fluxes, noise_variances, valid_masks,
                          lya_1pzs, z_qsos,
                          num_forest_lines, TW, OS,
                          dM_accum, dlog_omega_accum):
    """Loop over all spectra; accumulate gradients on .grad of params.
    Bypasses the buggy v1 wrapper. Returns total nlog_p (scalar, detached)."""
    omega2 = torch.exp(2 * log_omega)
    c_0 = torch.exp(log_c_0)
    tau_0 = torch.exp(log_tau_0)
    beta = torch.exp(log_beta)

    total = torch.zeros((), dtype=DTYPE)
    dM_accum.zero_()
    dlog_omega_accum.zero_()
    dlog_c_0_acc = torch.zeros((), dtype=DTYPE)
    dlog_tau_0_acc = torch.zeros((), dtype=DTYPE)
    dlog_beta_acc = torch.zeros((), dtype=DTYPE)

    n = centered_fluxes.shape[0]
    for i in range(n):
        valid = valid_masks[i]
        if not valid.any():
            continue
        y_m = centered_fluxes[i, valid]
        nv_m = noise_variances[i, valid]
        lya_1pz_m = lya_1pzs[i, valid]
        M_m = M[valid, :]
        omega2_m = omega2[valid]
        zqso_1pz = torch.tensor(float(z_qsos[i]) + 1.0, dtype=DTYPE)

        nlog_p, dM, dlog_omega, dlog_c_0, dlog_tau_0, dlog_beta = spectrum_loss(
            y_m, lya_1pz_m, nv_m, M_m, omega2_m,
            c_0, tau_0, beta,
            num_forest_lines, TW, OS, zqso_1pz,
        )
        total = total + nlog_p.detach()
        dM_accum[valid, :] += dM.detach()
        dlog_omega_accum[valid] += dlog_omega.detach()
        dlog_c_0_acc = dlog_c_0_acc + dlog_c_0.detach()
        dlog_tau_0_acc = dlog_tau_0_acc + dlog_tau_0.detach()
        dlog_beta_acc = dlog_beta_acc + dlog_beta.detach()

    # Apply BOSS DR12Q priors on log_tau_0, log_beta
    dlog_tau_0_acc = dlog_tau_0_acc + tau_0 * (tau_0 - TAU_0_PRIOR_MU) / TAU_0_PRIOR_SIGMA**2
    dlog_beta_acc = dlog_beta_acc + beta * (beta - BETA_PRIOR_MU) / BETA_PRIOR_SIGMA**2

    return total, dlog_c_0_acc, dlog_tau_0_acc, dlog_beta_acc


def _train_lbfgs_lane(lane, n_iters, init, train, t0_global,
                      use_strong_wolfe=True, history_size=10):
    """Step A.3 4th lane: torch.optim.LBFGS with strong-Wolfe line search.

    Uses v1 spectrum_loss (approximate dlog_β) for an apples-to-apples
    comparison with MATLAB's minFunc/L-BFGS — same loss kernel, same
    optimizer family. If endpoints match MATLAB to a small tolerance,
    that validates torch.optim.LBFGS for this problem class.
    """
    spectrum_loss = _import_spectrum_loss("v1")
    print(f"\n=== lane: {lane} (torch.optim.LBFGS, "
          f"line_search={'strong_wolfe' if use_strong_wolfe else 'None'}, "
          f"history={history_size}, n_iters={n_iters}) ===")

    mu, M, log_omega, log_c_0, log_tau_0, log_beta, TW, OS, num_forest_lines = \
        _build_init_params(init)

    centered_fluxes = torch.tensor(train["centered_fluxes"], dtype=DTYPE)
    centered_fluxes = torch.where(torch.isfinite(centered_fluxes), centered_fluxes,
                                  torch.zeros_like(centered_fluxes))
    noise_variances = torch.tensor(train["noise_variances"], dtype=DTYPE)
    valid_masks = torch.tensor(train["valid_masks"], dtype=torch.bool)
    z_qsos = train["z_qsos"]
    rest_wavelengths = init["rest_wavelengths"]
    lya_1pzs = _compute_lya_1pzs(rest_wavelengths, z_qsos)

    optimizer = torch.optim.LBFGS(
        [M, log_omega, log_c_0, log_tau_0, log_beta],
        lr=1.0,                # line search picks step internally
        max_iter=1,            # one L-BFGS step per .step() call so we can log
        history_size=history_size,
        line_search_fn="strong_wolfe" if use_strong_wolfe else None,
        tolerance_grad=0.0,    # we control termination ourselves
        tolerance_change=0.0,
    )

    history = dict(loss=[], log_c_0=[], log_tau_0=[], log_beta=[], wall_s=[])
    dM_accum = torch.zeros_like(M)
    dlog_omega_accum = torch.zeros_like(log_omega)
    last_loss = [None]   # closure-captured

    def closure():
        # 1) zero existing .grad on params
        optimizer.zero_grad()
        # 2) compute loss + manually-set gradients (mirrors v1 pattern)
        total, dlog_c_0_acc, dlog_tau_0_acc, dlog_beta_acc = _full_batch_objective(
            spectrum_loss, mu, M, log_omega, log_c_0, log_tau_0, log_beta,
            centered_fluxes, noise_variances, valid_masks,
            lya_1pzs, z_qsos, num_forest_lines, TW, OS,
            dM_accum, dlog_omega_accum,
        )
        # 3) populate .grad (L-BFGS reads .grad after closure() returns)
        with torch.no_grad():
            M.grad = dM_accum.clone()
            log_omega.grad = dlog_omega_accum.clone()
            log_c_0.grad = dlog_c_0_acc.clone()
            log_tau_0.grad = dlog_tau_0_acc.clone()
            log_beta.grad = dlog_beta_acc.clone()
        # 4) return loss as a tensor — L-BFGS uses it for line-search comparison
        loss_val = float(total)
        last_loss[0] = loss_val
        return torch.tensor(loss_val, dtype=DTYPE)

    for it in range(n_iters):
        t0 = time.time()
        optimizer.step(closure)
        dt = time.time() - t0
        history["loss"].append(last_loss[0])
        history["log_c_0"].append(float(log_c_0.detach()))
        history["log_tau_0"].append(float(log_tau_0.detach()))
        history["log_beta"].append(float(log_beta.detach()))
        history["wall_s"].append(float(dt))
        if it < 5 or it % 5 == 0 or it == n_iters - 1:
            wall = time.time() - t0_global
            print(f"  it={it:>3d}  loss={last_loss[0]:>14.4f}  "
                  f"τ_0={float(torch.exp(log_tau_0.detach())):.6f}  "
                  f"β={float(torch.exp(log_beta.detach())):.4f}  "
                  f"c_0={float(torch.exp(log_c_0.detach())):.6f}  "
                  f"({dt:.2f}s/iter, total {wall:.0f}s)")

    OUT.mkdir(parents=True, exist_ok=True)
    out = OUT / f"{lane}.npz"
    np.savez(out,
        lane=lane, n_iters=n_iters, lr=1.0,
        line_search="strong_wolfe" if use_strong_wolfe else "none",
        history_size=history_size,
        loss_history=np.asarray(history["loss"]),
        log_c_0_history=np.asarray(history["log_c_0"]),
        log_tau_0_history=np.asarray(history["log_tau_0"]),
        log_beta_history=np.asarray(history["log_beta"]),
        wall_s_history=np.asarray(history["wall_s"]),
        M_final=M.detach().numpy(),
        mu=mu.numpy(),
        log_omega_final=log_omega.detach().numpy(),
        log_c_0_final=float(log_c_0.detach()),
        log_tau_0_final=float(log_tau_0.detach()),
        log_beta_final=float(log_beta.detach()),
        c_0_final=float(torch.exp(log_c_0.detach())),
        tau_0_final=float(torch.exp(log_tau_0.detach())),
        beta_final=float(torch.exp(log_beta.detach())),
        rest_wavelengths=init["rest_wavelengths"],
    )
    print(f"  [saved] {out}")


def _train_one_lane(lane, n_iters, lr, init, train, t0_global):
    spectrum_loss = _import_spectrum_loss(lane)
    print(f"\n=== lane: {lane} (Adam, lr={lr}, n_iters={n_iters}) ===")

    mu, M, log_omega, log_c_0, log_tau_0, log_beta, TW, OS, num_forest_lines = \
        _build_init_params(init)

    centered_fluxes = torch.tensor(train["centered_fluxes"], dtype=DTYPE)
    centered_fluxes = torch.where(torch.isfinite(centered_fluxes), centered_fluxes, torch.zeros_like(centered_fluxes))
    noise_variances = torch.tensor(train["noise_variances"], dtype=DTYPE)
    valid_masks = torch.tensor(train["valid_masks"], dtype=torch.bool)
    z_qsos = train["z_qsos"]
    rest_wavelengths = init["rest_wavelengths"]
    lya_1pzs = _compute_lya_1pzs(rest_wavelengths, z_qsos)

    optimizer = torch.optim.Adam([M, log_omega, log_c_0, log_tau_0, log_beta], lr=lr)

    history = dict(loss=[], log_c_0=[], log_tau_0=[], log_beta=[], wall_s=[])
    dM_accum = torch.zeros_like(M)
    dlog_omega_accum = torch.zeros_like(log_omega)

    for it in range(n_iters):
        t0 = time.time()
        optimizer.zero_grad()
        total, dlog_c_0_acc, dlog_tau_0_acc, dlog_beta_acc = _full_batch_objective(
            spectrum_loss, mu, M, log_omega, log_c_0, log_tau_0, log_beta,
            centered_fluxes, noise_variances, valid_masks,
            lya_1pzs, z_qsos, num_forest_lines, TW, OS,
            dM_accum, dlog_omega_accum,
        )
        # Set .grad manually (mirrors v1 pattern; loss is detached)
        M.grad = dM_accum.clone()
        log_omega.grad = dlog_omega_accum.clone()
        log_c_0.grad = dlog_c_0_acc
        log_tau_0.grad = dlog_tau_0_acc
        log_beta.grad = dlog_beta_acc

        optimizer.step()
        dt = time.time() - t0

        history["loss"].append(float(total))
        history["log_c_0"].append(float(log_c_0.detach()))
        history["log_tau_0"].append(float(log_tau_0.detach()))
        history["log_beta"].append(float(log_beta.detach()))
        history["wall_s"].append(float(dt))

        if it % 5 == 0 or it == n_iters - 1:
            wall = time.time() - t0_global
            print(f"  it={it:>3d}  loss={float(total):>14.4f}  "
                  f"τ_0={float(torch.exp(log_tau_0)):.6f}  "
                  f"β={float(torch.exp(log_beta)):.4f}  "
                  f"c_0={float(torch.exp(log_c_0)):.6f}  "
                  f"({dt:.2f}s/iter, total {wall:.0f}s)")

    OUT.mkdir(parents=True, exist_ok=True)
    out = OUT / f"{lane}.npz"
    np.savez(out,
        lane=lane, n_iters=n_iters, lr=lr,
        loss_history=np.asarray(history["loss"]),
        log_c_0_history=np.asarray(history["log_c_0"]),
        log_tau_0_history=np.asarray(history["log_tau_0"]),
        log_beta_history=np.asarray(history["log_beta"]),
        wall_s_history=np.asarray(history["wall_s"]),
        M_final=M.detach().numpy(),
        mu=mu.numpy(),
        log_omega_final=log_omega.detach().numpy(),
        log_c_0_final=float(log_c_0.detach()),
        log_tau_0_final=float(log_tau_0.detach()),
        log_beta_final=float(log_beta.detach()),
        c_0_final=float(torch.exp(log_c_0.detach())),
        tau_0_final=float(torch.exp(log_tau_0.detach())),
        beta_final=float(torch.exp(log_beta.detach())),
        rest_wavelengths=init["rest_wavelengths"],
    )
    print(f"  [saved] {out}")


def main():
    p = argparse.ArgumentParser()
    p.add_argument("--lane", choices=["v1", "v3.5", "lbfgs", "lbfgs-no-ls", "both"], default="both")
    p.add_argument("--n-iters", type=int, default=50)
    p.add_argument("--lr", type=float, default=0.01)
    args = p.parse_args()

    init, train = _load_fixture()
    print(f"  loaded {train['centered_fluxes'].shape[0]} spectra "
          f"× {train['centered_fluxes'].shape[1]} pixels")
    print(f"  init c_0={float(init['c_0']):.4f}  tau_0={float(init['tau_0']):.5f}  "
          f"beta={float(init['beta']):.4f}")

    t0_global = time.time()
    if args.lane in ("v1", "both"):
        _train_one_lane("v1", args.n_iters, args.lr, init, train, t0_global)
    if args.lane in ("v3.5", "both"):
        _train_one_lane("v3.5", args.n_iters, args.lr, init, train, t0_global)
    if args.lane == "lbfgs":
        _train_lbfgs_lane("lbfgs", args.n_iters, init, train, t0_global,
                          use_strong_wolfe=True)
    if args.lane == "lbfgs-no-ls":
        # Diagnostic — torch L-BFGS WITHOUT line search (the "doesn't work"
        # configuration). For comparison.
        _train_lbfgs_lane("lbfgs-no-ls", args.n_iters, init, train, t0_global,
                          use_strong_wolfe=False)
    print(f"\n  total wall time: {time.time() - t0_global:.0f}s")


if __name__ == "__main__":
    main()
